spec/settings.md

```diff
@@ -117,22 +117,61 @@ settings. Other clients should respect this and not show the prompt if the user
 allowed the widget.
 
 For each room, the widget IDs that are allowed to load are stored in the event defined below.
-The widget's ID is the object's key, with the value being whether or not that widget ID is
-allowed to load. If a widget's ID is not in this event, it should be assumed as *not* allowed
+The widget's ID and URL is the object's key, with the value being whether or not that widget ID
+is allowed to load. If a widget's ID is not in this event, it should be assumed as *not* allowed
 to load (ie: `false`).
 
+For the widget URL in the key: it should not be set with the template filled out. It is the
+literal URL as defined by the widget.
+
+Account/user widgets do not need to use this prompt.
+
 **Event type**: `im.vector.setting.allowed_widgets`
 
 **Levels**: Only `room-account`
 
 **Content**:
 ```json
 {
-    "<widget ID>": true
+    "<widget ID>_<widget URL>": true
 }
 ```
 
 **Default**:
 ```json
 {}
 ```
+
+**Example**:
+
+A widget which looks like:
+```json
+{
+  "origin_server_ts": 1543854381750,
+  "sender": "@alice:example.org",
+  "event_id": "$1543854381213sKqbg:example.org",
+  "state_key": "customwidget_%40alice%3Aexample.org_1543007630924",
+  "content": {
+    "url": "https://scalar.vector.im/api/widgets/generic.html?url=$curl&room_id=$matrix_room_id",
+    "type": "customwidget",
+    "data": {
+        "curl": "https://matrix.org"
+    },
+    "name": "Custom",
+    "id": "customwidget_%40alice%3Aexample.org_1543007630924",
+    "waitForIframeLoad": true,
+    "creatorUserId": null
+  },
+  "type": "im.vector.modular.widgets",
+  "room_id": "!somewhere:example.org"
+}
+```
+
+would result in the following room account data `content`:
+```json
+{
+    "customwidget_%40alice%3Aexample.org_1543007630924_https://scalar.vector.im/api/widgets/generic.html?url=$curl&room_id=$matrix_room_id": true
```

Comment:
Did you consider something like:

```
{
"customwidget_%40alice%3Aexample.org_1543007630924": "c420e6fe0b017a22768f0a44f4beddcf8e6e2793"
}
```

Where c420e6fe0b017a22768f0a44f4beddcf8e6e2793 is shasum({the widget object}) ?

This would allow us at some point in future say that the widget needs new permission because it's changed (rather than just presenting it as a whole new widget).

Reply:
I'd rather not have all the Riot clients agree on a canonical json implementation in order to achieve that stably. 

The substantial part of the widget is the URL - the rest is purely aesthetic.

Reply:
A canonical json implementation of the setting? Or the widget?

_Is_ the URL the only material part? If we're not including the event ID or the room ID what stops somebody transporting an approved widget to a room in which the approval is dangerous? Is the 'data' not also material?

Reply:
To get consistent hashes we'd need to format the object the same way. Canonical JSON is how we do this elsewhere.

The room ID is implicit by nature of this being a room account data setting.

Using the event ID is probably fine, but still annoying to look up. It'll also result in more warnings as people only change titles. We also don't tell people what `data` contains because it's hard-coded values that usually end up in the URL. The personal information variables are not included in `data`.

Reply:
Ah - do we not just serve the JSON as it is written in the database? I thought it wouldn't matter if the clients wrote the JSON differently as long as it could be read "as it was written", but if that's not possible it's probably not worth the faff.

Personal information isn't included in data, but if the data represents the configuraiton of the widget and if the configuration of the widget could result in personal data being shared/accessible in a different scope then I think it's worth including in the representation of the widget's identity.

Reply:
> The room ID is implicit by nature of this being a room account data setting.

Understood

Reply:
> Ah - do we not just serve the JSON as it is written in the database?

JSON keys have no specific ordering rules. Even with Java/Kotlin the keys will end up in a different order from JavaScript. This is where canonical JSON fixes the problem.

> ...if the configuration of the widget could result in personal data being shared/accessible in a different scope...

It can't. The personal information (avatar, room ID, etc) is passed as spec-defined variables. The other variables are for the purposes of clients which natively render the widgets (which by the way might be another question) - the widget itself is kept the same.

Reply:
> JSON keys have no specific ordering rules.

Understood.

> It can't.

I was looking at the example and the data section looked like: 
```
    "data": {
        "curl": "https://matrix.org"
    },
```
Which made me think that even if the data you're sending to the widget isn't changing, if via the `data` section you could materially alter the widget's behaviour, your data _could_ end up being shared in a different way.

If the `data` section doesn't work that way then fine :)

Reply:
fuuuuuuuuuuuu-

yes, that is how it works and I've been wrong this whole time. I forgot that we put URLs and crap in there :/

I guess we just track the event ID and accept that the user will be extra prompted sometimes?

Reply:
> I guess we just track the event ID and accept that the user will be extra prompted sometimes?

Yeah, sometimes it'll be a cosmetic change that triggers a prompt; I think that's fine for now (better to fail safe, at least).